---
title: "Gene expression changes following forced promoter methylation"
author: "Keegan Korthauer"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
   html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Summary

Code to analyse the RNA-Seq and WGBS counts from the paper by Ford et al., 
"Frequent lack of repressive capacity of promoter DNA methylation identified 
through genome-wide epigenomic manipulation" (
available on [BioRxiv](https://www.biorxiv.org/content/early/2017/09/20/170506)).

Here we want to answer the question: is gene expression repressed after
forced promoter DNA methylation?

# Workspace Setup

First we load the necessary packages.

```{r, workspace, message=FALSE, warning=FALSE}
library(GEOquery)
library(data.table)
library(dplyr)
library(tidyr)
library(GenomicRanges)
library(ggplot2)
library(R.utils)
library(DESeq2)
library(annotatr)
library(bsseq)
library(dmrseq)
library(BiocParallel)
library(chunked)
library(DelayedMatrixStats)

# data directory
datdir <- "../DATA"
dir.create(datdir, showWarnings = FALSE)

# rscript directory
rdir <- "../R/"
sourceDirectory(rdir)

# results directory
resdir <- "../RESULTS"
dir.create(resdir, showWarnings = FALSE)

# parallel backend
ncores <- 6
register(MulticoreParam(workers=ncores))

# FDR significance levels for differential expression and DMR analyses
fdrlevel.dmr <- 0.10
fdrlevel.de <- 0.05
```

# Data download

The WGBS methylation counts for all cytosines are provided on GEO 
(accession number GSE102395). 
We will download the cytosine methylation counts for the following WGBS samples:

>GSM2735962 	MCF7_emptyVector_doxInduced_rep1_WGBS
>GSM2735963 	MCF7_empytVector_doxWithdrawn_rep1_WGBS
>GSM2735964 	MCF7_ZF_DNMT3A_doxInduced_rep1_WGBS
>GSM2735965 	MCF7_ZF_DNMT3A_doxInduced_rep2_WGBS
>GSM2735966 	MCF7_ZF_DNMT3A_doxInduced_rep3_WGBS
>GSM2735967 	MCF7_ZF_DNMT3A_doxWithdrawn_rep1_WGBS
>GSM2735968 	MCF7_ZF_DNMT3A_doxWithdrawn_rep2_WGBS
>GSM2735969 	MCF7_ZF_DNMT3A_noDox_rep1_WGBS

As well as the expression counts. First we get the cytosine methylation counts.
Since we are only interested in the CpG methylation context, we also filter 
these files by rows where the 4th column is equal to `CG`. This step requires that
the system has the `awk` command available.

```{r, downloadGEO}
# methylation counts
meta = pData(phenoData(getGEO("GSE102395")[[1]]))
filePaths <- as.character(meta$supplementary_file_1[grepl("WGBS", meta$description)])
corrupt <- NULL

for (file in filePaths){
  fi <- file.path(datdir,
              sapply(strsplit(file, "/"), function(x) x[length(x)]))
  err <- gsub(".gz", "", fi)
  if (!file.exists(fi) && !file.exists(gsub(".gz", "", fi))){
    download.file(url = file, destfile = fi)
  }
   
  if (!file.exists(gsub(".gz", "", fi))){
    err <- try(gunzip(fi), silent = TRUE) 
  }
  
  if (err==gsub(".gz", "", fi)){
    message("Downloaded and unzipped file ", fi)
  
    if (!file.exists(gsub("allC", "CpG", gsub(".gz", "", fi)))){
      system(paste0("awk '($4 == \"CG\")' ", gsub(".gz", "", fi),
                  " > ", gsub("allC", "CpG", gsub(".gz", "", fi))))
    }
    message("CpGs filtered for file ", gsub("allC", "CpG", gsub(".gz", "", fi)))
  }else{
    message(err)
    message("Skipping file ", fi, " since it is corrupt.")
    corrupt <- c(corrupt, fi)
    file.remove(fi)
  }
}
```

Since one of the files failed download, manually download from Lister lab's
public website (per email with Ethan Ford on 3/21/2018).

```{r, downloadFailedFile}
if (length(corrupt) == 1){
  if (!file.exists(corrupt) && !file.exists(gsub(".gz", "", corrupt))){
     download.file(url = "http://cpebrazor.ivec.org/public/listerlab/ethan/Public/MCF7_ZF_DNMT3A_doxInduced_rep2.allC.gz",
       destfile = corrupt)
  } 
  err <- gsub(".gz", "", corrupt)  
  if (!file.exists(gsub(".gz", "", corrupt))){
     err <- try(gunzip(corrupt), silent = TRUE) 
  }
  
  if (err==gsub(".gz", "", corrupt)){
    message("Downloaded and unzipped file ", corrupt)
  
    if (!file.exists(gsub("allC", "CpG", gsub(".gz", "", corrupt)))){
      system(paste0("awk '($4 == \"CG\")' ", gsub(".gz", "", corrupt),
                  " > ", gsub("allC", "CpG", gsub(".gz", "", corrupt))))
    }
    message("CpGs filtered for file ", gsub("allC", "CpG", gsub(".gz", "", corrupt)))
  }
}else if (length(corrupt) > 1){
  stop("More than one file is corrupt")
}else{
  message("All files downloaded and unzipped")
}
```

Next, we download the expression count matrix.

```{r, downloadGEO2}
# expression counts
file <- file.path(datdir, "GSE102395_MCF7_ZF_DNMT3A_countstable.txt")

if (!file.exists(file)){
  download.file(url = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE102nnn/GSE102395/suppl/GSE102395%5FMCF7%5FZF%5FDNMT3A%5Fcountstable%2Etxt%2Egz",
              destfile = paste0(file, ".gz"),
              mode="wb")
  gunzip(paste0(file, ".gz"))
}
```

# Read in data tables

Now we'll read in the methylation count tables and convert them to bsseq objects.
This uses the `read.lister` function sourced in by the script in the `R` directory.

```{r, read.lister}
bfile <- file.path(datdir, "bsseq.rds")

if (!file.exists(bfile)){
  files <- list.files(datdir, pattern = "rep[[:digit:]].CpG.txt", full.names = TRUE)
  
  bs <- read.lister(files, sampleNames = files, rmZeroCov = TRUE,
                    strandCollapse = TRUE, mc.cores = ncores)
  if ("BSseq" %in% class(bs)){
    saveRDS(bs, bfile)
  }
}else{
  bs <- readRDS(bfile)
}

show(bs)
```

We'll add the metadata so we can easily access the condition information in 
further analyses.

```{r, meta}
if(ncol(pData(bs))==0){
  sampleNames(bs) <- unlist(sapply(strsplit(sampleNames(bs), "../DATA/"), 
                                   function(x) x[[2]]))
  sampleNames(bs) <- unlist(sapply(strsplit(sampleNames(bs), ".CpG"), 
                                 function(x) x[[1]]))
  meta$basefile <- unlist(sapply(strsplit(as.character(meta$supplementary_file_1),
                                          "suppl/"), function(x) x[[length(x)]]))
  meta$basefile <- unlist(sapply(strsplit(meta$basefile, ".allC"), 
                                 function(x) x[[1]]))
  
  pdat <- meta[meta$basefile %in% sampleNames(bs),] %>%
    mutate(condition = ifelse(grepl("Vector|noDox", source_name_ch1), 
                              "Control", "Methylated"),
           dox = ifelse(grepl("noDox", source_name_ch1), "No DOX", 
                        ifelse(grepl("doxInduced", source_name_ch1), "DOX", 
                               "Dox Withdrawal")))
  pData(bs) <- pdat
  saveRDS(bs, bfile)
}
```

## Exploratory analysis

Here will look at some sample similarity metrics among the samples to verify that
control samples are most similar to controls and ZF dox samples are more similar to 
other ZF dox samples (this is indeed the case from the coorelation patrix and 
clustering dendrogram below).

```{r}
cov.mat <- getCoverage(bs, type="Cov")
filter <- pmax( 1*(rowSums2(cov.mat[,pData(bs)$condition == "Control"]) >= 5),
                1*(rowSums2(cov.mat[,pData(bs)$condition == "Methylated"]) >= 5))
filter <- which(filter > 0)
bs.filt <- bs[-filter,]
rm(cov.mat)

cormat <- round(cor(as.matrix(getMeth(bs.filt, type="raw")),
                    use = "pairwise.complete.obs", 
                    method = "spearman"),2)
rownames(cormat) <- colnames(cormat) <- labs <- paste0(pData(bs)$condition, "_", 
                                               pData(bs)$dox, 
                                               "_Sample", 1:ncol(cormat))
cormat <- data.frame(cormat) %>%
  mutate(Sample1 = labs) %>%
  gather("Sample2", "Correlation", 1:ncol(cormat))

cormat$Sample1 <- factor(cormat$Sample1) 
cormat$Sample2 <- factor(cormat$Sample2)

ggplot(data = cormat, aes(x=Sample1, y=Sample2, fill=Correlation)) + 
  geom_tile() +
  scale_fill_gradient(low="white", high="red") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

d <- hclust(dist(t(as.matrix(as.data.frame(getMeth(bs.filt, type="raw")))),
                 method="euclidean"))
cols = c(rep("green", 2), rep("red", 3), rep("purple", 2), "blue")
plot(d, labels = rep("", length(labs)), xlab="")
mtext(labs, side = 1, line = c(-1, -7, -14, -14, -19, -8, -14.5, -1), 
      at = order(d$order),
        col = cols, las = 2, cex = 0.9)
rm(bs.filt)
```

First we filter on the samples we will use for the analysis. Specifically, we'll
keep only the control samples and the DOX-induced samples with the MCF7-ZF vector.

```{r, filtsamples}
bs <- bs[,pData(bs)$condition == "Control" | 
          pData(bs)$dox == "DOX"]
```

Here, we'll look at weighted smoothed density plots of coverage and methylation 
proportion. Plots of coverage by group and sample:

```{r, methplot, fig.width = 8, fig.height = 6}
set.seed(486)
idx <- sample(1:nrow(bs), 1e6)
plotEmpiricalDistribution(bs[idx,], 
                          bySample = TRUE,
                          testCovariate = "condition",
                          type = "Cov")
```

Plots of beta values by group and sample:

```{r, methplot2, fig.width = 8, fig.height = 6}
plotEmpiricalDistribution(bs[idx,], 
                          bySample = TRUE,
                          testCovariate = "condition",
                          adj = 3)
```


# DMR analysis

```{r, eval=FALSE, results='hide', echo=FALSE, include=FALSE}
# test out if parallel is functioning properly
library(BiocParallel)
tf <- function(){
  message("a.", appendLF=FALSE)
  Sys.sleep(1)
}

#without parallel
system.time({a <- lapply(1:16, function(x) tf())})

register(MulticoreParam(8))  
system.time({a <- bplapply(1:16, function(x) tf())})
```

We will use dmrseq to identify Differentially Methylated Regions (DMRs). 

First we will filter the loci by coverage (remove those with zero coverage
in at least one condition).

```{r, filter}
cov.mat <- getCoverage(bs, type="Cov")
filter <- pmax( 1*(rowSums2(cov.mat[,pData(bs)$condition == "Control"]) == 0),
                1*(rowSums2(cov.mat[,pData(bs)$condition == "Methylated"]) == 0))
filter <- which(filter > 0)
bs <- bs[-filter,]
rm(cov.mat)
```

This removed `r length(filter)` out of `r nrow(bs)` loci (
`r signif(100*length(filter)/nrow(bs), 2)`%). Now we'll run dmrseq.

```{r, dmrseq, message=FALSE}
resfile <- file.path(resdir, "regions.rds")

if (!file.exists(resfile)){
  regions <- dmrseq(bs, testCovariate = "condition",
                    bpSpan = 500,
                    maxGap = 500,
                    maxPerms = 6,
                    chrsPerChunk = 2)
  saveRDS(regions, file=resfile)
}else{
  regions <- readRDS(resfile)
}

sum(regions$qval < fdrlevel.dmr)
```

We'll add the raw mean methylation differences to the region summaries and plot
the top DMR.

```{r, fig.width = 15, fig.height= 5}
regions$meanDiff <- meanDiff(bs, dmrs=regions, testCovariate="condition")

plotDMRs(bs, regions=regions[1,], testCovariate="condition")
```

Next, we'll look at the overlap of the dmrseq DMRs with the DSS DMRs provided by
the authors.

```{r, compare}
dssfile <- file.path(datdir, "170506-2.txt")
if (!file.exists(dssfile)){
  download.file(url = "https://www.biorxiv.org/highwire/filestream/57972/field_highwire_adjunct_files/1/170506-2.txt",
              destfile =  dssfile)
}

dmrs.dss <- fread(dssfile, header=FALSE)  #DMRs
colnames(dmrs.dss) <- c("chr", "start", "end", "number_of_CpG_cytosines_in_region",
                   "methylation_level_in_MCF7_control", "methylation_level_in_ZF_D3A_plus_dox",
                   "methylation_level_in_ZF_D3A_dox_wd", "closest_gene_promoter",
                   "distance_to_nearest_promoter", "promoter_classification",
                   "CpGisland_classification", "genebody_classification",
                   "enhancer_classification")

dmrs.dss <- makeGRangesFromDataFrame(dmrs.dss, keep.extra.columns = TRUE)
dmrs.dss$meanDiff <- meanDiff.dss(bs, dmrs=dmrs.dss, testCovariate="condition")

ol <- findOverlaps(regions[regions$qval < fdrlevel.dmr, ], dmrs.dss)
```

It looks like `r length(unique(ol@from))` 
(`r signif(100*(length(unique(ol@from)) / length(dmrs.dss)), 3)`%) 
of the DSS DMRs overlap with the dmrseq DMRs (at `r fdrlevel.dmr` FDR). 
Conversely, `r length(unique(ol@to))` 
(`r signif(100*(length(unique(ol@to)) / sum(regions$qval < fdrlevel.dmr)), 3)`%)

A couple of plots to compare the characteristics of regions found by DSS and 
dmrseq.

```{r, fig.width= 12, fig.heigh=12}
par(mfrow=c(2,2))
plot(density(log2(dmrs.dss$number_of_CpG_cytosines_in_region+1)),
     main = "Smoothed density of number of CpGs",
     xlab = "Number of CpGs per region (log2+1)")
lines(density(log2(regions[regions$qval < fdrlevel.dmr, ]$L+1)), col="blue")

plot(density(log2(width(dmrs.dss)+1)),
     main = "Smoothed density of region width",
     xlab = "Number of basepairs per region (log2+1)")
lines(density(log2(width(regions[regions$qval < fdrlevel.dmr, ])+1)), col="blue")

legend(11, 0.4, legend=c("DSS", "dmrseq"), col=c("black", "blue"), lty=c(1,1))

plot(density((regions[regions$qval < fdrlevel.dmr, ]$L /
                     width(regions[regions$qval < fdrlevel.dmr, ]))), 
     col="blue",
     main = "Smoothed density of CpG density",
     xlab = "Number of CpGs per basepair (log2+1)")
lines(density((dmrs.dss$number_of_CpG_cytosines_in_region/width(dmrs.dss))))

plot(density(dmrs.dss$meanDiff),
     main = "Smoothed density of meanDiff",
     xlab = "Mean proportion difference in methylation")
lines(density(regions[regions$qval < fdrlevel.dmr, ]$meanDiff),col="blue")

```

# DE analysis

First we need to run the differential expression analysis. 

```{r, expressionEDA, fig.width=8, fig.height=4}
exp <- fread(file.path(datdir, "GSE102395_MCF7_ZF_DNMT3A_countstable.txt"))

# There are duplicate genes -- looks like a classic excel conversion error -- remove these
dup <- which(table(exp$gene) > 1)
names(dup)
exp <- exp %>% filter(!(gene %in% names(dup)))

# move genes to rownames  
rownames(exp) <- exp$gene
exp <- as.matrix(exp %>% dplyr::select(-gene))
allZero <- rowSums(exp==0)==ncol(exp)
exp <- exp[!allZero,]

# strings to match the two comparison samples - control and ZF + DOX
ctrl <- "MCF7_emptyVector|MCF7_ZF_DNMT3A_noDOX_rep"
trt <- "MCF7_ZF_DNMT3A_DOX"

# set up colData
coldata <- data.frame(sample=colnames(exp)) %>%
  mutate(condition = ifelse(grepl(ctrl, sample), "Control",
                     ifelse(grepl(trt, sample), "Methylated", "Other"))) 

dds <- DESeqDataSetFromMatrix(countData = exp,
                              colData = coldata,
                              design = ~ condition)

# plot distribution of expression values
exp <- data.frame(counts(dds)) %>% 
  mutate(gene = rownames(dds)) %>%
  gather(sample, count, 1:19) %>% 
  mutate(condition = ifelse(grepl(ctrl, sample), "Control", "Methylated")) %>%
  mutate(dox = ifelse(grepl("noDOX", sample), "No Dox", 
                      ifelse(grepl("DOXremoval", sample), "Dox withdrawal", 
                             "Dox")))
  
ggplot(exp, aes(x=count+1, group=sample, color=condition)) +
  geom_line(stat="density", adjust=0.75, lwd=1, aes(linetype=dox), alpha=0.5) +
  scale_x_continuous(trans="log2") +
  theme_bw() + 
  ggtitle("Smoothed density of raw counts")

# collapse over samples of the same dox/condition 
ggplot(exp, aes(x=count+1, group=interaction(dox, condition), color=condition)) +
  geom_line(stat = "density", adjust=0.75, aes(linetype=dox), alpha = 0.5, lwd=1) +
  scale_x_continuous(trans="log2") +
  theme_bw() + 
  ggtitle("Smoothed density of raw counts")

dds <- dds[, !( dds$condition == "Methylated" & grepl("removal", dds$sample))]
dds$condition <- droplevels(dds$condition)
```

Since we observe global shifts in the expected
direction (i.e. lower expression for the methylated condition), we will not perform
standard normalization. Instead, we'll compute size factors only on genes that
were at least 10kb away from a ZF binding site or putative DMR (by dmrseq, with
FDR level 0.25).
To find these genes, we'll use the ZF ChIP-seq binding
site data from the Supplementary table S1.

```{r, norm}
download.file(url = "https://www.biorxiv.org/highwire/filestream/57972/field_highwire_adjunct_files/0/170506-1.txt",
              destfile =  file.path(datdir, "170506-1.txt"))

ts1 <- fread(file.path(datdir, "170506-1.txt"), header=FALSE, skip=1)  #ZF Binding sites
colnames(ts1) <- c("chr", "start", "end", "closest_gene_promoter", 
                   "distance_to_nearest_promoter", "promoter_classification",
                   "CpGisland_classification", "genebody_classification",
                   "enhancer_classification")

annot = build_annotations(genome = 'hg19', annotations = 'hg19_genes_promoters')
ol <- distanceToNearest(regions[regions$qval < 0.25,], annot)
ol2 <- findOverlaps(regions[regions$qval < 0.25,], annot)
dmr_genes <- unique(c(annot$symbol[ol@to[mcols(ol)$distance <= 10000]],
                      annot$symbol[ol2@to]))
dmr_genes <- dmr_genes[!is.na(dmr_genes)]

zf_genes <- (ts1 %>%
  filter(distance_to_nearest_promoter <= 10000))$closest_gene_promoter

norm_genes <- unique((exp %>%
  filter(!(gene %in% unique(c(dmr_genes, zf_genes)))))$gene)

exp <- exp %>% 
  mutate(normg = ifelse(gene %in% norm_genes, "Background Gene", "ZF-bound Gene"))

ggplot(exp %>% filter(dox != "Dox withdrawal"), 
       aes(x=count+1, group=interaction(normg, condition), color=normg)) +
  geom_line(stat = "density", alpha =0.5, adjust=0.75, lwd=1, aes(linetype=condition)) +
  scale_x_continuous(trans="log2") +
  theme_bw() + 
  ggtitle("Smoothed density of raw counts") 
```

Looks like the ZF-bound genes in the Dox group had slightly shifted global expression.
For this reason, we only compute sizefactors on the normgenes.
We also filter out very lowly expressed genes (total counts in all samples together
fewer than 10 or counts of zero in more than half of the samples). 
We'll use the 4 no dox control versus the 4 dox methylated group for the DE comparison. 

```{r, deseq}
dds <- dds[, dds$condition == "Methylated" | 
             (dds$condition == "Control" & grepl("noDOX", dds$sample)) ]
dds <- dds[rowSums(counts(dds)) >= 10,]
dds <- dds[rowSums(counts(dds)==0) <= 0.5,]
dds <- estimateSizeFactors(dds, 
                           controlGenes = rownames(dds) %in% norm_genes)
dds_global <- estimateSizeFactors(dds)
table(dds$condition)

# check resulting distributions after normalization
exp <- data.frame(counts(dds)) %>% 
  mutate(gene = rownames(dds)) %>%
  gather(sample, count, 1:ncol(dds)) %>% 
  mutate(count = (data.frame(counts(dds, normalize=FALSE)) %>% 
         gather(sample, count))$count) %>%
  mutate(normcount = (data.frame(counts(dds, normalize=TRUE)) %>% 
         gather(sample, count))$count) %>%
  mutate(normcount_global = (data.frame(counts(dds_global, normalize=TRUE)) %>% 
         gather(sample, count))$count) %>%
  mutate(condition = ifelse(grepl(ctrl, sample), "no-dox", "+dox")) %>%
  mutate(dox = ifelse(grepl("noDOX", sample), "No Dox","Dox"))

pgen <- ggplot(exp, aes(x=count, group=sample, color=condition)) +
  scale_x_continuous(trans="log1p", breaks=c(1,100,1000,1000,100000)) +
  theme_bw() +
  scale_color_manual(values = c("red", "blue")) +
  labs(color = "Treatment") 

praw <- pgen + 
  xlab("Raw count") +
  ggtitle("Smoothed density of raw counts") +
  geom_line(aes(x=count, group=sample, color=condition),
            stat="density", alpha = 0.4, lwd=1, adjust=0.75) 

pglobal <- pgen +
  geom_line(aes(x=normcount_global, group=sample, color=condition),
            stat="density", alpha = 0.4, lwd=1, adjust=0.75) +
  xlab("Globally normalized count") +
  ggtitle("Smoothed density of globally normalized counts") 

pcontrol <- pgen +
  geom_line(aes(x=normcount, group=sample, color=condition),
            stat="density", alpha = 0.4, lwd=1, adjust=0.75) +
  xlab("Control-gene normalized count") +
  ggtitle("Smoothed density of control-gene normalized counts") 

praw
pglobal
pcontrol

saveRDS(praw,file="../plots/expdist_raw.rds")
saveRDS(pglobal,file="../plots/expdist_global.rds")
saveRDS(pcontrol,file="../plots/expdist_control.rds")

dds <- DESeq(dds)
res <- results(dds)

sum(res$padj < fdrlevel.de, na.rm = TRUE)
dds <- dds[!is.na(res$padj)]
res <- res %>% na.omit()
```

In contrast, here are the results using global normalization:

```{r}
dds_global <- DESeq(dds_global)
sum(results(dds_global)$padj < fdrlevel.de, na.rm = TRUE)
```

# Association with effect size

Here we'll create several figures exploring the relationship between methylation and expression,
using effect size (methylation proportion difference) as the statistic of interest to 
rank DMRs.

##  Associate dmrseq DMRs with genes

Next we need to associate each DMR with a gene by checking overlap with promoters.
We'll use the promoter annotation from annotatr.

```{r, annotatr-dmrseq}
regions.sig <- regions[regions$qval < fdrlevel.dmr,]

ol <- distanceToNearest(regions.sig, annot)

dmrs <- regions.sig[ol@from[mcols(ol)$distance <= 2000],]
dmrs$gene <- annot$symbol[ol@to[mcols(ol)$distance <= 2000]]
dmrs$prom <- annot[ol@to[mcols(ol)$distance <= 2000],]
  
dmrs <- dmrs[!is.na(dmrs$gene),]
```

Now that we have a set of DMR-Gene associations, we'll add the the DE information 
to the DMR data frame.

```{r}
# add DE results
baseMeanPerLvl <- sapply( levels(dds$condition), 
                          function(lvl) 
                            rowMeans(counts(dds,normalized=TRUE)[,dds$condition == lvl] ))
res <- cbind(res, baseMeanPerLvl)
x <- match(dmrs$gene, rownames(res)) 
dmrs <- dmrs[!is.na(x),]
x <- x[!is.na(x)]
res.dmrs <- res[x,]

dmrs <- cbind(DataFrame(dmrs), res.dmrs)
```

Now annotate with CG Island status.

```{r}
annot_islands = build_annotations(genome = 'hg19', 
                                  annotations = 'hg19_cpg_islands')
dmrs$CGI <- countOverlaps(dmrs$prom, annot_islands, minoverlap = 100) > 0

```

Prepare a figure that stratifies counts of DMR-DE pairs by CGI status and 
direction of effect. Do this only for DMRs reaching 0.01 cutoff.

```{r}
dmrs01 <- data.frame(dmrs) %>% filter(qval < 0.01)
tab.CGI <- table(dmrs01$CGI[dmrs01$padj < fdrlevel.de], 
                 dmrs01$log2FoldChange[dmrs01$padj < fdrlevel.de] < 0)
chisq.test(tab.CGI)
tab.tot <- table(dmrs01$log2FoldChange[dmrs01$padj < fdrlevel.de] < 0)

tab.CGI <- data.frame(count = c(as.vector(tab.CGI),
                                as.vector(tab.tot)),
                      prop = c(as.vector(tab.CGI/t(matrix(rep(colSums(tab.CGI),2), ncol=2))),
                                as.vector(tab.tot/sum(tab.tot))),
                      CGI = c("non-CGI", "CGI", "non-CGI", "CGI", "All", "All"),
                      Genes = c("Activated", "Activated", 
                                "Repressed", "Repressed",
                                "Activated", "Repressed"))

tab.CGI$CGI <- factor(tab.CGI$CGI, levels=c("All", "non-CGI", "CGI"))
tab.CGI$Genes <- factor(tab.CGI$Genes, levels=c("Activated", "Repressed"))

pb <- ggplot(tab.CGI, aes(y = count, x = CGI, fill = Genes)) +
  geom_bar(stat = "identity", position = "dodge") +
  xlab("CG density of promoter") +
  ylab("Count") + 
  scale_fill_manual(values = c("#E69F00", "#0072B2")) +
  labs(fill = "Differential\nExpression")+
  theme_bw()
pb
saveRDS(pb, file = "../plots/barplot_CGI.rds")
```

## Scatterplot of methylation versus expression

Now we create the scatterplot of methylation versus expression for dmrs.

```{r, dmrs-5a-dmrseq}
dmrsa <- data.frame(dmrs@listData) %>% 
  mutate(sig = padj < fdrlevel.de,
         FC = 2^log2FoldChange,
         delta_mC = meanDiff) %>%
  mutate(A = 0.5*(log2(baseMean) + 
               log2(FC*baseMean))) %>%
  mutate(ABin = ntile(A, 9)) %>%
  filter(!is.na(log2FoldChange) & !is.na(sig)) 

ggplot(dmrsa, aes(x = delta_mC, y = log2FoldChange)) +
  geom_hline(yintercept=0, col="black") +
  geom_hline(yintercept=0, col="white", linetype="dashed") +
  geom_point(size=0.5, alpha=0.75, aes(color = sig)) + 
  theme_bw() + 
  xlab(expression(paste(Delta, "mCG in DMR"))) +
  ylab("log2 fold change mRNA abundance") +
  scale_color_manual(values=c("black", "red")) +
  geom_smooth() +
  labs(color="Differentially\n Expressed")  +
  geom_point(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred") +
  geom_label(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred",
            aes(label="SOX2"), hjust=0.5, vjust=1.25) +
  ggtitle("dmrseq DMRs")
```

From this figure it is a little more clear that most DE genes are down-regulated.

## Histogram of the observed expression fold changes in DMRs

First in the style of the paper, for easy comparison:
```{r, dmrs-5b-orig-dmrseq, fig.height=4, fig.width=7}
dmrsb <- dmrsa %>% 
  filter(delta_mC > 0.3) %>%
  mutate(FCcat = cut(FC,
                     breaks=c(0,0.3,0.5,0.7,0.9,1.1,Inf))) %>%
  mutate(color = ifelse(FCcat %in% c("(0.9,1.1]", "(1.1,Inf]"), "Increase", 
                        ifelse(FCcat == "(0.7,0.9]", "Small Decrease", "Decrease"))) %>%
  mutate(FCcat = factor(FCcat, levels = rev(levels(FCcat))),
         sig = ifelse(padj < fdrlevel.de, "Differentially Expressed", 
                      "Not Differentially Expressed"))

ggplot(dmrsb, aes(FCcat, fill = color)) +
  geom_bar() + 
  theme_bw() + 
  scale_fill_manual(values=c("red", "blue", "darkgreen")) + 
  xlab("mRNA fold change") +
  ylab("Number of genes") +
  ggtitle("All Gene - dmrseq DMR Pairs") + 
  annotate("text", x = 6, y = 800, label = paste0("n=",nrow(dmrsb))) +
  labs(fill="Direction \nof effect")
```

How many genes had a decrease compared to an increase overall?

```{r}
de_dn <- sum(dmrsb$log2FoldChange < 0, na.rm=TRUE)
de_up <- sum(dmrsb$log2FoldChange > 0, na.rm=TRUE)
```
Looks like there are `r de_dn` decreased genes and `r de_up` increased genes, where
the proportion decreased is `r signif(de_dn / (de_dn + de_up),3)` and the odds of decrease is
`r signif(de_dn / de_up,3)`.

Next in the improved symmetrical style and separating by DE status:
```{r,fig.height=5, fig.width=10} 
dmrsc <- dmrsa %>% 
  filter(delta_mC > 0.3) %>%
  mutate(sig = ifelse(padj < fdrlevel.de, "Differentially Expressed", 
                      "Not Differentially Expressed")) %>%
  mutate(color = ifelse(log2FoldChange > 0, "Increase", "Decrease")) %>%
  mutate(FCcat = cut(log2FoldChange, breaks = c(-6,-4,-2,-1,-0.5,0,0.5,1,2,4,6))) %>%
  na.omit()

ggplot(dmrsc , aes(FCcat, fill = color)) +
  geom_bar() + 
  facet_wrap( ~ sig) +
  geom_vline(xintercept = 4.5, linetype = "dashed") +
  theme_bw() + 
  scale_fill_manual(values=c("red", "blue", "darkgreen")) + 
  xlab("mRNA fold change") + 
  ylab("Number of genes") +
  ggtitle("Gene - dmrseq DMR Pairs by DE significance") +
  labs(fill="Direction \nof effect")
```

How many DE genes had a decrease compared to an increase overall?

```{r}
de_dn <- sum(dmrsc$sig == "Differentially Expressed" & dmrsc$log2FoldChange < 0, na.rm=TRUE)
de_up <- sum(dmrsc$sig == "Differentially Expressed" & dmrsc$log2FoldChange > 0, na.rm=TRUE)
```
Looks like there are `r de_dn` decreased DE genes and `r de_up` increased DE genes, where
the proportion decreased is `r signif(de_dn / (de_dn + de_up),3)` and the odds of decrease is
`r signif(de_dn / de_up,3)`.

# Association with dmrseq statistic

Here we'll create several figures exploring the relationship between methylation and expression,
using the dmrseq statistic as the statistic of interest to 
rank DMRs. This should be more robust than ordering regions simply on their average methylation difference.

## Scatterplot of methylation versus expression

Now we create the scatterplot of methylation and expression, but replacing the x-axis 
with the region-level test statistic instead of the methylation difference 
estimate.

```{r, dmrs-5a-dmrseq-alt}
dmrsa <- data.frame(dmrs@listData) %>% 
  mutate(sig = padj < fdrlevel.de,
         FC = 2^log2FoldChange,
         delta_mC = meanDiff) %>%
  mutate(A = 0.5*(log2(baseMean) + 
               log2(FC*baseMean))) %>%
  mutate(ABin = ntile(A, 9)) %>%
  mutate(ptcol2 = ifelse(sig, "Differentially Expressed", 
                         "Not Differentially Expressed"),
         ptcol3 = ifelse(qval > 0.01, "Not significant (DMR)", 
                        ifelse(sig, "DMR and DE", "Not significant (DE)"))) %>%
  filter(!is.na(log2FoldChange) & !is.na(sig)) 

ggplot(dmrsa, aes(x = stat, y = log2FoldChange)) +
  geom_hline(yintercept=0, col="black") +
  geom_hline(yintercept=0, col="white", linetype="dashed") +
  geom_point(size=0.5, alpha=0.75, aes(color = ptcol2)) + 
  theme_bw() + 
  xlab("Region test statistic") +
  ylab("log2 fold change mRNA abundance") +
  scale_color_manual(values=c("red", "black")) +
  geom_smooth() +
  labs(color="Significance")  +
  geom_point(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred") +
  geom_label(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred",
            aes(label="SOX2"), hjust=0.5, vjust=1.25) +
  ggtitle("dmrseq DMRs") +
  geom_vline(xintercept=min(dmrsa$stat[dmrs$qval<0.01]), 
             linetype="dashed", color="grey20")
```

Save these plots (all together and stratified by CGI) for plotting in main figures.

```{r, extraplots}

A2 <- ggplot(dmrsa, aes(x = stat, y = log2FoldChange)) +
  geom_hline(yintercept=0, col="black") +
  geom_hline(yintercept=0, col="white", linetype="dashed") +
  geom_point(size=0.5, alpha=0.75, aes(color = ptcol3)) + 
  theme_bw() + 
  xlab("Region test statistic") +
  ylab("log2 fold change mRNA abundance") +
  scale_color_manual(values=c("red", "black", "grey")) +
  geom_smooth() +
  labs(color="Significance")  +
  geom_point(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred") +
  geom_label(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred",
            aes(label="SOX2"), hjust=0.5, vjust=1.25) +
  ggtitle("dmrseq DMRs") +
  ylim(-3,3) + 
  geom_vline(xintercept=min(dmrsa$stat[dmrs$qval<0.01]), 
             linetype="dashed", color="grey20")
A2 

# save plot for mutlipanel figure
saveRDS(A2, file="../plots/A2.rds")


# Just CGI
A2 <- ggplot(dmrsa %>% filter(CGI), aes(x = stat, y = log2FoldChange)) +
  geom_hline(yintercept=0, col="black") +
  geom_hline(yintercept=0, col="white", linetype="dashed") +
  geom_point(size=0.5, alpha=0.75, aes(color = ptcol3)) + 
  theme_bw() + 
  xlab("Region test statistic") +
  ylab("log2 fold change mRNA abundance") +
  scale_color_manual(values=c("red", "black", "grey")) +
  geom_smooth() +
  labs(color="Significance")  +
  geom_point(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred") +
  geom_label(data =  dmrsa[dmrsa$gene == "SOX2",], col="darkred",
            aes(label="SOX2"), hjust=0.5, vjust=1.25) +
  ggtitle("dmrseq DMRs") +
  ylim(-3,3) + 
  geom_vline(xintercept=min(dmrsa$stat[dmrs$qval<0.01]), 
             linetype="dashed", color="grey20")
A2 

# save plot for mutlipanel figure
saveRDS(A2, file="../plots/A2_cgi.rds")

# Just nonCGI
A2 <- ggplot(dmrsa %>% filter(!CGI), aes(x = stat, y = log2FoldChange)) +
  geom_hline(yintercept=0, col="black") +
  geom_hline(yintercept=0, col="white", linetype="dashed") +
  geom_point(size=0.5, alpha=0.75, aes(color = ptcol3)) + 
  theme_bw() + 
  xlab("Region test statistic") +
  ylab("log2 fold change mRNA abundance") +
  scale_color_manual(values=c("red", "black", "grey")) +
  geom_smooth() +
  labs(color="Significance")  +
  ggtitle("dmrseq DMRs") +
  ylim(-3,3) + 
  geom_vline(xintercept=min(dmrsa$stat[dmrs$qval<0.01]), 
             linetype="dashed", color="grey20")
A2 

# save plot for mutlipanel figure
saveRDS(A2, file="../plots/A2_noncgi.rds")
```

From this figure it is much more clear that most DE genes are down-regulated.

## Histogram of the observed expression fold changes in DMRs

First in the style of the paper for easy comparison:
```{r, dmrs-5b-orig-dmrseq-alt, fig.height=4, fig.width=7}
dmrsb <- dmrsa %>% 
  filter(qval < 0.01) %>%
  mutate(FCcat = cut(FC,
                     breaks=c(0,0.3,0.5,0.7,0.9,1.1,Inf))) %>%
  mutate(color = ifelse(FCcat %in% c("(0.9,1.1]", "(1.1,Inf]"), "Increase", 
                        ifelse(FCcat == "(0.7,0.9]", "Small Decrease", "Decrease"))) %>%
  mutate(FCcat = factor(FCcat, levels = rev(levels(FCcat))))

ggplot(dmrsb, aes(FCcat, fill = color)) +
  geom_bar() + 
  theme_bw() + 
  scale_fill_manual(values=c("red", "blue", "darkgreen")) + 
  xlab("mRNA fold change") +
  ylab("Number of genes") +
  ggtitle("All Gene - dmrseq DMR Pairs") + 
  annotate("text", x = 6, y = 400, label = paste0("n=",nrow(dmrsb))) +
  labs(fill="Direction \nof effect")
```

How many genes had a decrease compared to an increase overall?

```{r}
de_dn <- sum(dmrsb$log2FoldChange < 0, na.rm=TRUE)
de_up <- sum(dmrsb$log2FoldChange > 0, na.rm=TRUE)
```
Looks like there are `r de_dn` decreased genes and `r de_up` increased genes, where
the proportion decreased is `r signif(de_dn / (de_dn + de_up),3)` and the odds of decrease is
`r signif(de_dn / de_up,3)`.

Next in the improved symmetrical style and separating by DE status:
```{r,fig.height=5, fig.width=10} 
dmrsc <- dmrsa %>%
  filter(qval < 0.01) %>%
  mutate(sig = ifelse(padj < fdrlevel.de, "Differentially Expressed", 
                      "Not Differentially Expressed")) %>%
  mutate(color = ifelse(log2FoldChange > 0, "Increase", "Decrease")) %>%
  mutate(FCcat = cut(log2FoldChange, breaks = c(-6,-4,-2,-1,-0.5,0,0.5,1,2,4,6))) %>%
  na.omit()

ggplot(dmrsc, aes(FCcat, fill = color)) +
  geom_bar() + 
  facet_wrap( ~ sig) +
  theme_bw() + 
  scale_fill_manual(values=c("red", "blue", "darkgreen")) + 
  xlab("mRNA fold change") + 
  ylab("Number of genes") +
  ggtitle("Gene - dmrseq DMR Pairs by DE significance") +
  labs(fill="Direction \nof effect")

```

Save these plots (all together and stratified by CGI) for plotting in main figures.

```{r, extraplots2}
B2 <- ggplot(dmrsc %>% 
               filter(sig == "Differentially Expressed"),
             aes(x=log2FoldChange)) +
  geom_histogram(breaks=seq(-3,3, by=1/5), 
                 aes(fill=color)) + 
  scale_fill_manual(values=c("red", "blue")) + 
  theme_bw() + 
  xlab("log2 mRNA fold change") + 
  ylab("Number of genes") +
  ggtitle("Gene - dmrseq DMR Pairs by DE significance") +
  labs(fill="Direction \nof effect") +
  xlim(-3,3) +
  ylim(0,329)
B2
# save plot for mutlipanel figure
saveRDS(B2, file="../plots/B2.rds")

# just CGI
B2 <- ggplot(dmrsc %>% 
               filter(sig == "Differentially Expressed") %>%
               filter(CGI),
             aes(x=log2FoldChange)) +
  geom_histogram(breaks=seq(-3,3, by=1/5), 
                 aes(fill=color)) + 
  scale_fill_manual(values=c("red", "blue")) + 
  theme_bw() + 
  xlab("log2 mRNA fold change") + 
  ylab("Number of genes") +
  ggtitle("CG Island Promoters") +
  labs(fill="Direction \nof effect") +
  xlim(-3,3) 
B2
# save plot for mutlipanel figure
saveRDS(B2, file="../plots/B2_cgi.rds")

# just nonCGI
B2 <- ggplot(dmrsc %>% 
               filter(sig == "Differentially Expressed") %>%
               filter(!CGI),
             aes(x=log2FoldChange)) +
  geom_histogram(breaks=seq(-3,3, by=1/5), 
                 aes(fill=color)) + 
  scale_fill_manual(values=c("red", "blue")) + 
  theme_bw() + 
  xlab("log2 mRNA fold change") + 
  ylab("Number of genes") +
  ggtitle("Non-CG Island Promoters") +
  labs(fill="Direction \nof effect") +
  xlim(-3,3) 
B2
# save plot for mutlipanel figure
saveRDS(B2, file="../plots/B2_noncgi.rds")
```

How many DE genes had a decrease compared to an increase overall?

```{r}
de_dn <- sum(dmrsc$sig == "Differentially Expressed" & dmrsc$log2FoldChange < 0, na.rm=TRUE)
de_up <- sum(dmrsc$sig == "Differentially Expressed" & dmrsc$log2FoldChange > 0, na.rm=TRUE)
```
Looks like there are `r de_dn` decreased DE genes and `r de_up` increased DE genes, where
the proportion decreased is `r signif(de_dn / (de_dn + de_up),3)` and the odds of decrease is
`r signif(de_dn / de_up,3)`.

The confidence intervals for these quantities are: 

```{r}
n <- de_dn + de_up
prop <- de_dn / n
odds <- de_dn / de_up

prop.ci <- prop + c(-1,1)*qnorm(0.975)*sqrt(prop*(1-prop)/n)
odds.ci <- prop.ci*n / ((1-prop.ci)*n)

prop.ci
odds.ci
```

This shows that DE genes are **overwhelmingly** 
downregulated when their promoter is demethylated.

# Session Information

```{r}
sessionInfo()
```
